flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/MessageAcknowledingSourceBase.java

```diff
@@ -157,17 +179,35 @@ public void restoreState(SerializedCheckpointData[] state) throws Exception {
 
 	@Override
 	public void notifyCheckpointComplete(long checkpointId) throws Exception {
-		for (Iterator<Tuple2<Long, List<Id>>> iter = pendingCheckpoints.iterator(); iter.hasNext();) {
-			Tuple2<Long, List<Id>> checkpoint = iter.next();
-			long id = checkpoint.f0;
-			
-			if (id <= checkpointId) {
-				acknowledgeIDs(checkpoint.f1);
-				iter.remove();
-			}
-			else {
-				break;
+		if (!running) {
+			LOG.debug("notifyCheckpointComplete() called on closed source");
+			return;
+		}
+		
+		// only one commit operation must be in progress
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("Committing Messages externally for checkpoint {}", checkpointId);
+		}
+		
+		List<Id> idsOfCheckpointedMessages = new LinkedList<Id>();
+		synchronized (pendingCheckpoints) {
+			for (Iterator<Tuple2<Long, List<Id>>> iter = pendingCheckpoints.iterator(); iter.hasNext();) {
+				Tuple2<Long, List<Id>> checkpoint = iter.next();
+				long id = checkpoint.f0;
+				
+				if (id <= checkpointId) {
+					idsForCurrentCheckpoint.addAll(checkpoint.f1);
+					acknowledgeIDs(checkpoint.f1);
+					iter.remove();
+				}
+				else {
+					break;
+				}
 			}
 		}
+		
+		if (LOG.isInfoEnabled()) {
+			LOG.info("Committing Messages with following IDs {}", idsOfCheckpointedMessages);
```

Comment:
This will in most cases generate a gigantic log line that may bring the whole system to a stall.